lib/assert_eventually.ex

```diff
@@ -123,15 +148,63 @@ defmodule AssertEventually do
               Process.sleep(@assert_eventually_interval)
               f.(f, start_time)
           end
+        else
+          unquote(assert)
         end
       end
 
       # pass defined function as argument to itself to allow for recursion
       fun.(fun, unquote(now_ts()))
-      unquote(assert)
     end
   end
 
+  # lists
+  defp neutralize_variable(list) when is_list(list) do
+    Enum.map(list, &neutralize_variable/1)
+  end
+
+  # lists with head separator
+  defp neutralize_variable({:|, meta, values}) do
+    {:|, meta, Enum.map(values, &neutralize_variable/1)}
+  end
+
+  # tuples
+  defp neutralize_variable({:{}, meta, rest}) do
+    {:{}, meta, Enum.map(rest, &neutralize_variable/1)}
+  end
+
+  # Exception - 2-elem tuples are represented as-is in AST
+  defp neutralize_variable({part1, part2}) do
+    {neutralize_variable(part1), neutralize_variable(part2)}
+  end
+
+  # maps
+  defp neutralize_variable({:%{}, meta, content}) do
+    {:%{}, meta,
+     Enum.map(content, fn
+       {key, variable} -> {key, neutralize_variable(variable)}
+     end)}
+  end
+
+  # structs
+  defp neutralize_variable({:%, meta, [aliases, map]}) do
+    {:%, meta, [aliases, neutralize_variable(map)]}
+  end
+
+  # variables
+  defp neutralize_variable({atom, meta, module}) when is_atom(atom) do
+    string_value = Atom.to_string(atom)
+
+    if String.starts_with?(string_value, "_") do
+      {atom, meta, module}
+    else
+      {:"_#{string_value}", meta, module}
+    end
+  end
+
+  # base values
+  defp neutralize_variable(variable), do: variable
```

Comment:
I'm a bit worried about this part - if this handles all the cases. Honestly - this probably doesn't handle all the cases, but we can always add them if we approach them. Also I'm not super confident about AST representation of structs - I think they are always a list of two tuples - one with `:__aliases__` and the other with the map that is underneath, but I couldn't find anything in docs to confirm that.

Reply:
This is also why I initially leaned towards suppressing warnings here instead if possible. But I think this is fine - if this is a breaking change it's gonna be possible to pin the older version until specific issue is resolved.